qsc/write_vmec_input.py

```diff
@@ -0,0 +1,283 @@
+          # Note: fzero returns its answer in phi0_rootSolve_min
+          phi0_solution = res.root
+
+          final_R, final_z = Frenet_to_cylindrical_1_point(phi0_solution)
+          R_2D[j_theta,j_phi] = final_R
+          z_2D[j_theta,j_phi] = final_z
+          phi0_2D[j_theta,j_phi] = phi0_solution
+
+    # Fourier transform the result.
+    # This is not a rate-limiting step, so for clarity of code, we don't bother with an FFT.
+    mpol = int(min(N_theta          / 2, mpol1d))
+    ntor = int(min(N_phi_conversion / 2, ntord))
+    mpol_nonzero = mpol
+    RBC = np.zeros((int(2*ntor+1),int(mpol+1)))
+    RBS = np.zeros((int(2*ntor+1),int(mpol+1)))
+    ZBC = np.zeros((int(2*ntor+1),int(mpol+1)))
+    ZBS = np.zeros((int(2*ntor+1),int(mpol+1)))
+    factor = 2 / (N_theta * N_phi_conversion)
+    for j_phi in range(N_phi_conversion):
+        for j_theta in range(N_theta):
+            for m in range(mpol):
+                nmin = -ntor
+                if m==0: nmin = 1
+                for n in range(nmin, ntor):
+                    angle = m * theta[j_theta] - n * self.nfp * phi_conversion[j_phi]
+                    sinangle = np.sin(angle)
+                    cosangle = np.cos(angle)
+                    factor2 = factor
+                    # The next 2 lines ensure inverse Fourier transform(Fourier transform) = identity
+                    if np.mod(N_theta,2) == 0 and m  == (N_theta/2): factor2 = factor2 / 2
+                    if np.mod(N_phi_conversion,2) == 0 and abs(n) == (N_phi_conversion/2): factor2 = factor2 / 2
+                    RBC[n,m] = RBC[n,m] + R_2D[j_theta, j_phi] * cosangle * factor2
+                    RBS[n,m] = RBS[n,m] + R_2D[j_theta, j_phi] * sinangle * factor2
+                    ZBC[n,m] = ZBC[n,m] + z_2D[j_theta, j_phi] * cosangle * factor2
+                    ZBS[n,m] = ZBS[n,m] + z_2D[j_theta, j_phi] * sinangle * factor2
+    RBC[0,0] = np.sum(R_2D) / (N_theta * N_phi_conversion)
+    ZBC[0,0] = np.sum(z_2D) / (N_theta * N_phi_conversion)
+
+    if lasym == False:
+        RBS = 0
+        ZBC = 0
+
+    # Write to VMEC file
+    File_object = open(filename,"w+")
+    File_object.write("! This &INDATA namelist was generated by pyQSC\n")
+    if vmec_template_filename!=[]:    File_object.write("! Based on template file ",vmec_template_filename+'\n')
+    File_object.write("! r ="+str(r)+'\n')
+    File_object.write('!----- Runtime Parameters -----\n')
+    File_object.write('&INDATA\n')
+    File_object.write('  DELT = '+str(delt)+'\n')
+    File_object.write('  NITER = '+str(niter)+'\n')
+    File_object.write('  NSTEP = '+str(nstep)+'\n')
+    File_object.write('  TCON0 = '+str(tcon0)+'\n')
+    File_object.write('  NS_ARRAY = '+str(ns_array)[1:-1]+'\n')
+    File_object.write('  FTOL_ARRAY = '+str(ftol_array)[1:-1]+'\n')
+    File_object.write('  NITER_ARRAY = '+str(niter_array)[1:-1]+'\n')
+    File_object.write('  FTOL_ARRAY = '+str(ftol_array)[1:-1]+'\n')
```

Comment:
Why is `[1:-1]` specified? This causes the first element to be skipped, and don't you want to keep the first element? (Also `[1:]` has the same effect as `[1:-1]`) And ftol_array is written twice.

Reply:
The [1:-1] command acts on the string. It simply removes the initial '[' and final ']' characters so that they are not outputted to the VMEC input file.